File: test/clients/ts/protocol-restart.mjs
/**
 * Real browser-process restart test.
 *
 * Chromium process 1 queues into native IndexedDB and commits to PostgreSQL,
 * then exits without acknowledging. Process 2 launches with the same browser
 * profile, reconstructs the exact SDK envelope from IndexedDB, receives the
 * durable duplicate result, and acknowledges. Process 3 proves that the local
 * acknowledgement itself survived another browser process boundary.
 */
import assert from 'node:assert/strict';
import { resolve } from 'node:path';

import { chromium } from '../../../../opto-sync-clients/clients/ts/node_modules/playwright/index.mjs';
import {
  bundleBrowserClient,
  serveBundle,
} from '../../../../opto-sync-clients/clients/ts/test/helpers/bundle.mjs';

const [profileArgument, recordId] = process.argv.slice(2);
if (!profileArgument || !recordId) {
  console.error('usage: node protocol-restart.mjs <browser-profile-dir> <record-id>');
  process.exit(2);
}

const profileDir = resolve(profileArgument);
const serverUrl = (
  process.env.OPTO_SYNC_SERVER_URL || 'http://localhost:3003'
).replace(/\/$/, '');
const databaseName = `opto-browser-restart-${recordId}`;
const HTML =
  '<!doctype html><meta charset="utf-8">' +
  '<title>opto-sync restart</title><script src="/opto-sync.browser.js"></script>';

await bundleBrowserClient();
const pageServer = await serveBundle(HTML);

async function launchPhase(run) {
  const context = await chromium.launchPersistentContext(profileDir, {
    headless: true,
  });
  try {
    const pages = context.pages();
    const page = pages[0] ?? (await context.newPage());
    await page.goto(`${pageServer.origin}/`, { waitUntil: 'load' });
    return await run(page);
  } finally {
    await context.close();
  }
}

try {
  const firstState = await launchPhase((page) =>
    page.evaluate(
      async ({ databaseName: dbName, recordId: id, serverUrl: base }) => {
        const { initOptoSync, OptoSyncClient } = window.OptoSync;
        await initOptoSync();
        const client = new OptoSyncClient({ databaseName: dbName });
        await client.queueMutation(
          'docs',
          id,
          { title: 'indexeddb survived server-commit/client-ack restart' },
          { baseRevision: '0' },
        );
        const envelope = await client.protocolPushRequest();
        const response = await fetch(`${base}/v1/sync/push`, {
          method: 'POST',
          headers: { 'content-type': 'application/json' },
          body: JSON.stringify(envelope),
        });
        const result = await response.json();
        if (!response.ok || result.results?.[0]?.status !== 'applied') {
          throw new Error(`initial push failed: ${response.status} ${JSON.stringify(result)}`);
        }
        const pending = await client.pendingMutations();
        if (pending.length !== 1) {
          throw new Error(`prepare must leave one pending mutation, got ${pending.length}`);
        }

        const snapshotResponse = await fetch(`${base}/v1/sync/snapshot`);
        const snapshot = await snapshotResponse.json();
        if (!snapshotResponse.ok) {
          throw new Error(
            `snapshot failed: ${snapshotResponse.status} ${JSON.stringify(snapshot)}`,
          );
        }
        let interrupted = false;
        try {
          await client.installSnapshot(snapshot, async () => {
            localStorage.setItem('restart-authoritative', 'partial');
            throw new Error('injected snapshot replacement interruption');
          });
        } catch (error) {
          if (!String(error).includes('injected snapshot replacement interruption')) throw error;
          interrupted = true;
        }
        if (
          !interrupted ||
          (await client.pullCheckpoint()) !== '0' ||
          (await client.pendingMutations()).length !== 1
        ) {
          throw new Error('interrupted snapshot advanced checkpoint or changed pending work');
        }
        client.db.close();
        return { envelope, snapshot };
      },
      { databaseName, recordId, serverUrl },
    ),
  );
  console.log(
    'ok - [ts/restart prepare] Chromium committed without local ack; ' +
      'snapshot replacement interrupted',
  );

  await launchPhase((page) =>
    page.evaluate(
      async ({ databaseName: dbName, original, snapshot, serverUrl: base }) => {
        const { initOptoSync, OptoSyncClient } = window.OptoSync;
        await initOptoSync();
        const client = new OptoSyncClient({ databaseName: dbName });
        const pending = await client.pendingMutations();
        if (pending.length !== 1) {
          throw new Error(`restart recovered ${pending.length} pending mutations, expected 1`);
        }
        if (
          (await client.pullCheckpoint()) !== '0' ||
          localStorage.getItem('restart-authoritative') !== 'partial'
        ) {
          throw new Error('new Chromium process did not observe interrupted snapshot state');
        }
        await client.installSnapshot(snapshot, async (records) => {
          localStorage.setItem('restart-authoritative', JSON.stringify(records));
        });
        if (
          (await client.pullCheckpoint()) !== snapshot.checkpoint ||
          (await client.pendingMutations()).length !== 1 ||
          localStorage.getItem('restart-authoritative') === 'partial'
        ) {
          throw new Error('snapshot retry did not repair authoritative state');
        }
        const reconstructed = await client.protocolPushRequest();
        if (JSON.stringify(reconstructed) !== JSON.stringify(original)) {
          throw new Error(
            `envelope changed across browser restart: ` +
              `${JSON.stringify(original)} != ${JSON.stringify(reconstructed)}`,
          );
        }
        const response = await fetch(`${base}/v1/sync/push`, {
          method: 'POST',
          headers: { 'content-type': 'application/json' },
          body: JSON.stringify(reconstructed),
        });
        const result = await response.json();
        const mutation = result.results?.[0];
        if (
          !response.ok ||
          mutation?.status !== 'duplicate' ||
          mutation?.originalStatus !== 'applied'
        ) {
          throw new Error(`retry was not deduplicated: ${response.status} ${JSON.stringify(result)}`);
        }
        const acknowledged = await client.acknowledgePush(
          result,
          reconstructed,
        );
        if (acknowledged !== 1 || (await client.pendingMutations()).length !== 0) {
          throw new Error('recovered acknowledgement did not drain exactly one mutation');
        }
        client.db.close();
      },
      {
        databaseName,
        original: firstState.envelope,
        snapshot: firstState.snapshot,
        serverUrl,
      },
    ),
  );
  console.log(
    'ok - [ts/restart recover] snapshot repaired; ' +
      'identical retry deduplicated and acknowledged',
  );

  await launchPhase((page) =>
    page.evaluate(async ({ databaseName: dbName, snapshotCheckpoint }) => {
      const { initOptoSync, OptoSyncClient } = window.OptoSync;
      await initOptoSync();
      const client = new OptoSyncClient({ databaseName: dbName });
      const pending = await client.pendingMutations();
      if (pending.length !== 0) {
        throw new Error(`acknowledged mutation returned after second restart: ${pending.length}`);
      }
      if ((await client.pullCheckpoint()) !== snapshotCheckpoint) {
        throw new Error('installed snapshot checkpoint did not survive second restart');
      }
      const authoritative = localStorage.getItem('restart-authoritative');
      if (!authoritative || authoritative === 'partial' || !Array.isArray(JSON.parse(authoritative))) {
        throw new Error('repaired authoritative snapshot did not survive second restart');
      }
      await client.db.delete();
      localStorage.removeItem('restart-authoritative');
    }, { databaseName, snapshotCheckpoint: firstState.snapshot.checkpoint }),
  );
  console.log(
    'ok - [ts/restart verify] snapshot checkpoint and acknowledgement survived ' +
      'a second Chromium process',
  );
  assert.ok(firstState.envelope.clientId, 'SDK envelope did not contain a clientId');
} finally {
  await pageServer.close();
}

File: test/clients/rust/src/bin/protocol_restart.rs
use std::fs;
use std::path::Path;

use opto_sync_client::protocol::{PushRequest, PushResponse, SnapshotResponse};
use opto_sync_client::protocol_sync::AtomicProtocolSyncStore;
use opto_sync_client::sqlite::SqliteProtocolStore;
use opto_sync_client_e2e::{protocol_push, protocol_snapshot};
use serde_json::{json, Value};

fn fail(message: impl AsRef<str>) -> ! {
    panic!("{}", message.as_ref())
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 5 || !matches!(args[1].as_str(), "prepare" | "recover") {
        eprintln!(
            "usage: protocol_restart <prepare|recover> <sqlite-file> <envelope-file> <record-id>"
        );
        std::process::exit(2);
    }
    let phase = &args[1];
    let state_path = Path::new(&args[2]);
    let envelope_path = Path::new(&args[3]);
    let snapshot_path = format!("{}.snapshot", envelope_path.display());
    let snapshot_path = Path::new(&snapshot_path);
    let record_id = &args[4];

    if phase == "prepare" {
        if state_path.exists() || envelope_path.exists() || snapshot_path.exists() {
            fail("prepare requires fresh state and envelope files");
        }
        let nonce = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let client_id = format!("rust-restart-{}-{nonce}", std::process::id());
        let mut store = SqliteProtocolStore::open(state_path, &client_id).unwrap();
        store
            .queue_upsert_record(
                "docs",
                record_id,
                json!({"title": "rust survived server-commit/client-ack restart"}),
                Some("0".to_string()),
                false,
            )
            .unwrap();
        let mut queue = store.load_queue().unwrap();
        let envelope = serde_json::to_value(queue.push_request(100).unwrap()).unwrap();
        fs::write(envelope_path, serde_json::to_vec(&envelope).unwrap()).unwrap();

        let committed = protocol_push(&envelope);
        assert_eq!(committed.status, 200, "{}", committed.body);
        assert_eq!(
            committed.json()["results"][0]["status"],
            "applied",
            "{}",
            committed.body
        );
        assert_eq!(queue.pending().count(), 1, "prepare must not acknowledge");

        let snapshot_result = protocol_snapshot();
        assert_eq!(snapshot_result.status, 200, "{}", snapshot_result.body);
        let snapshot: SnapshotResponse = serde_json::from_str(&snapshot_result.body).unwrap();
        fs::write(snapshot_path, &snapshot_result.body).unwrap();
        store
            .connection()
            .execute_batch(
                "CREATE TEMP TRIGGER inject_snapshot_failure
                 BEFORE INSERT ON _opto_sync_local
                 BEGIN
                   SELECT RAISE(ABORT, 'injected snapshot replacement interruption');
                 END;",
            )
            .unwrap();
        queue.set_checkpoint(snapshot.checkpoint.clone()).unwrap();
        let interrupted = store.replace_authoritative_and_persist(&snapshot.records, &mut queue);
        assert!(interrupted.is_err());
        let durable = store.load_queue().unwrap();
        assert_eq!(durable.checkpoint(), "0");
        assert_eq!(durable.pending().count(), 1);
        assert_eq!(
            store
                .local_record("docs", record_id)
                .unwrap()
                .unwrap()
                .record["title"],
            "rust survived server-commit/client-ack restart"
        );
        println!(
            "ok - [rust/sqlite restart prepare] committed without local ack; \
             SQLite snapshot transaction interrupted"
        );
        return;
    }

    let original: Value = serde_json::from_slice(&fs::read(envelope_path).unwrap()).unwrap();
    let request: PushRequest = serde_json::from_value(original.clone()).unwrap();
    let mut store = SqliteProtocolStore::open(state_path, &request.client_id).unwrap();
    let mut queue = store.load_queue().unwrap();
    assert_eq!(
        queue.pending().count(),
        1,
        "new process did not recover one pending mutation"
    );
    assert_eq!(queue.checkpoint(), "0");

    let snapshot: SnapshotResponse =
        serde_json::from_slice(&fs::read(snapshot_path).unwrap()).unwrap();
    queue.set_checkpoint(snapshot.checkpoint.clone()).unwrap();
    store
        .replace_authoritative_and_persist(&snapshot.records, &mut queue)
        .unwrap();
    assert_eq!(queue.checkpoint(), snapshot.checkpoint);
    assert_eq!(queue.pending().count(), 1);
    assert!(store
        .authoritative_record("docs", record_id)
        .unwrap()
        .is_some());

    let reconstructed = serde_json::to_value(queue.push_request(100).unwrap()).unwrap();
    assert_eq!(
        reconstructed, original,
        "SDK envelope changed across process restart"
    );

    let retry = protocol_push(&reconstructed);
    assert_eq!(retry.status, 200, "{}", retry.body);
    assert_eq!(retry.json()["results"][0]["status"], "duplicate");
    assert_eq!(retry.json()["results"][0]["originalStatus"], "applied");
    let acknowledgement: PushResponse = serde_json::from_str(&retry.body).unwrap();
    assert_eq!(queue.acknowledge(&acknowledgement, &request).unwrap(), 1);
    store
        .persist_acknowledgement(&mut queue, &request, &acknowledgement)
        .unwrap();
    assert_eq!(queue.pending().count(), 0);
    drop(store);
    let reopened = SqliteProtocolStore::open(state_path, &request.client_id).unwrap();
    assert_eq!(reopened.load_queue().unwrap().pending().count(), 0);
    println!(
        "ok - [rust/sqlite restart recover] snapshot repaired; \
         identical retry deduplicated and acknowledgement survived reopen"
    );
}
